src/lib/replication/repl_dev/raft_repl_dev.h

```diff
@@ -380,6 +383,25 @@ class RaftReplDev : public ReplDev,
     void reset_quorum_size(uint32_t commit_quorum);
     void create_snp_resync_data(raft_buf_ptr_t& data_out);
     bool save_snp_resync_data(nuraft::buffer& data, nuraft::snapshot& s);
+
+    ReplServiceError init_req_ctx(repl_req_ptr_t rreq, repl_key rkey, journal_type_t op_code, bool is_proposer,
+                                  sisl::blob const& user_header, sisl::blob const& key, uint32_t data_size,
+                                  cshared< ReplDevListener >& listener);
+
+    // no_space_left error handling
+    void pause_statemachine();
+    void resume_statemachine();
+    void set_no_space_left_error_info(repl_lsn_t lsn, chunk_num_t chunk_id);
+    void reset_no_space_left_error_info();
+    void handle_no_space_left();
+
+    void enter_emergency() { m_in_emergency.store(true, std::memory_order_release); }
+    void leave_emergency() { m_in_emergency.store(false, std::memory_order_release); }
+    bool is_in_emergency() { return m_in_emergency.load(std::memory_order_acquire); }
+
+    void inc_pending_init_req_num() { m_pending_init_req_num.fetch_add(1, std::memory_order_acq_rel); }
+    void dec_pending_init_req_num() { m_pending_init_req_num.fetch_sub(1, std::memory_order_acq_rel); }
+    uint64_t get_pending_init_req_num() { return m_pending_init_req_num.load(std::memory_order_acquire); }
```

Comment:
suggestion RAII pattern for these kind of pending_io counting
```
class OperationCounter {
public:
    OperationCounter() {
       counter.fetch_add(1)
    }

    ~OperationCounter() {
        counter.fetch_add(-1)
    }

    static int getCount() {
        return counter;
    }

private:
    static std::atomic_uint64_t counter;
};
```

Reply:
we can not use static here, since  OperationCounter will be share by diffferent raft_repl_dev, OperationCounter should be valid for a particular raft_dev_dev

Reply:
I think the key point is not the static,  but more about encapsulate the inc_pending and dec_pending into RAII, similar to cp_guard().    There are too many places that need to do dec_pending() ,  not only here,  the graceful-shutdown has more....  the dec spread across every `else` is a concerning.

Sorry this is a bit off-topic.

Reply:
done, change has been made @xiaoxichen , ptal